test/common/http/http1/codec_impl_test.cc

```diff
@@ -1257,31 +1255,61 @@ TEST_P(Http1ServerConnectionImplTest, HeaderInvalidAuthority) {
 // Mutate an HTTP GET with embedded NULs, this should always be rejected in some
 // way (not necessarily with "head value contains NUL" though).
 TEST_P(Http1ServerConnectionImplTest, HeaderMutateEmbeddedNul) {
-  if (parser_impl_ == ParserImpl::BalsaParser) {
-    // TODO(#21245): Re-enable this test for BalsaParser.
+  const absl::string_view example_input = "GET / HTTP/1.1\r\nHOST: h.com\r\nfoo: barbaz\r\n";
+
+  for (size_t n = 0; n < example_input.size(); ++n) {
+    initialize();
+
+    InSequence sequence;
+
+    MockRequestDecoder decoder;
+    EXPECT_CALL(callbacks_, newStream(_, _)).WillOnce(ReturnRef(decoder));
+
+    Buffer::OwnedImpl buffer(absl::StrCat(
+        example_input.substr(0, n), absl::string_view("\0", 1), example_input.substr(n),
+        // TODO(#21245): Fix BalsaParser to process headers before final "\r\n".
+        Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http1_use_balsa_parser") ? "\r\n"
+                                                                                           : ""));
+    EXPECT_CALL(decoder, sendLocalReply(_, _, _, _, _));
+    auto status = codec_->dispatch(buffer);
+    EXPECT_FALSE(status.ok()) << n;
+    EXPECT_TRUE(isCodecProtocolError(status));
+    EXPECT_THAT(status.message(), testing::HasSubstr("http/1.1 protocol error:"));
+  }
+}
+
+// Mutate the trailers with an HTTP POST with embedded NULs.
+// This should always be rejected.
+TEST_P(Http1ServerConnectionImplTest, TrailerMutateEmbeddedNul) {
+  if (parser_impl_ == ParserImpl::HttpParser) {
```

Comment:
Would this test not also work for HttpParser? If not, what's the reason? Seems like it might be valuable for both.

Reply:
It passes for http-parser for most values of `n` except for 6, 7, 21, 22, when the NUL character is injected around the two : characters.  Not sure why.

Reply:
I think that might be due to trailers being off unless `codec_settings_.enable_trailers_ = true;` is configured. Does it work if that's added at the start of the test?